src/firmware/app/control/trajectory_planner.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+#include "firmware/shared/math/polynomial_2d.h"
+#include "firmware/shared/math/vector_2d.h"
+
+#define TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS                                              \
+    6000  // The maximum size of the array containing trajectory elements
+#define __TRAJECTORY_PLANNER_MAX_PATH_LENGTH__ 18  // [meters]
+
+// Struct that defines a single point on a trajectory
+// Includes the Position,and Timedata corresponding to that point
+typedef struct TrajectoryElement
+{
+    Vector2d_t position;
+    double time;
+} TrajectoryElement_t;
+
+typedef struct Trajectory
+{
+    TrajectoryElement_t* trajectory_elements;
+    unsigned int num_elements;
+} Trajectory_t;
+
+typedef struct FirmwareRobotPathParameters
+{
+    Polynomial2dOrder3_t path;
+    double t_start;
+    double t_end;
+    unsigned int num_segments;
+    double max_allowable_acceleration;
+    double max_allowable_speed;
+    double initial_speed;
+    double final_speed;
+} FirmwareRobotPathParameters_t;
+
+/**
+ * Returns a planned trajectory as a list of TrajectoryElement's.
+ *
+ * Key assumptions of this planner are:
+ *  - Trajectories are time-optimal assuming INFINITE JERK capability of the robot
+ *  - No speed on the trajectory is larger than the 'max_allowable_speed' parameter.
+ *  - No acceleration value between points on the trajectory can be larger than the
+ * 'max_allowable_acceleration' input parameter
+ *  - Assuming the grip-limit of the robot IS THE SAME AS THE MAX ACCELERATION and at no
+ * point on the path can the sum of centripital and acceleration force be greather than
+ * 'max_allowable_acceleration'
+ *
+ *  Trajectory generation is done by assuming constant acceleration capability.
+ *      - The generator will assume maximum acceleration for the robot between each
+ * segment on the path. If using max acceleration breaches either the speed limit of the
+ * robot, or the centripital acceleration limit defind by the curvature and robot speed -
+ * the planner will assume the max acceleration that will remain bellow this limit.
+ *      - To ensure the robot is capable of deccelerating to reach ant sudden changes in
+ * the path, the trajectory is checked for backwards continuity
+ *
+ * @pre t_start < t_end
+ * @pre num_segments > 2
+ * @pre max_allowable_acceleration > 0
+ * @pre max_allowable_speed > 0
+ * @pre init_speed > 0
+ * @pre final_speed > 0
```

Comment:
We should add `assert`'s for all of these, as well as checking that `num_segments < TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS`

Reply:
Added asserts for the pre conditions